### PR TITLE
[tp] disable some assertion temporarily for torch.compile

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116678
* __->__ #116573
* #116559
* #116426

Disable some runtime assertion first as it does not work with
torch.compile properly, I'll have a follow up fix in dynamo and reenable
this check again

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/distributed/tensor/parallel/style.py [hunk 1]
```diff
@@ -280,12 +280,14 @@ def _prepare_input_fn(self, inputs, device_mesh):
         prepared_inputs = []
         if not isinstance(inputs, tuple):
             inputs = (inputs,)
-        assert len(inputs) == len(self.input_layouts), \
-            "module inputs and input_layouts should have same length!"
+        if len(inputs) != len(self.input_layouts):
+            raise RuntimeError("module inputs and input_layouts should have same length!")
```
Comment (Skylion007) [inline torch/distributed/tensor/parallel/style.py:284]:
```suggestion
            raise AssertionError("module inputs and input_layouts should have same length!")
```

File: torch/distributed/tensor/parallel/style.py [hunk 2]
```diff
@@ -354,12 +356,13 @@ def _prepare_out_fn(self, outputs, device_mesh):
         prepared_outputs = []
         if not isinstance(outputs, tuple):
             outputs = (outputs,)
-        assert len(outputs) == len(self.output_layouts), \
-            "module outputs and output_layouts should have same length!"
+        if len(outputs) != len(self.output_layouts):
+            raise RuntimeError("module outputs and output_layouts should have same length!")
```
Comment (Skylion007) [inline torch/distributed/tensor/parallel/style.py:360]:
```suggestion
            raise AssertionError("module outputs and output_layouts should have same length!")
```

File: torch/distributed/tensor/parallel/style.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/tensor/parallel/style.py]: updated thanks!

**Review by awgu — APPROVED** [review]
Sounds good!

(can consider filing an issue to track adding these back if you think it would help)

**Review by XilunWu — APPROVED** [review]
good to know that assertion causes trouble when interacting with torch.compile

**Comment by wanchaol** [comment]
> Sounds good!
> 
> (can consider filing an issue to track adding these back if you think it would help)

sounds good! filed a issue https://github.com/pytorch/pytorch/issues/116712

**Pull request closed without merging**